cpp/src/explainer/tree_shap.cu

```diff
@@ -0,0 +1,326 @@
+              gpu_treeshap::PathElement<MySplitCondition<ThresholdType>>{
+                path_idx,
+                tree.SplitIndex(parent_idx),
+                group_id,
+                {lower_bound, upper_bound},
+                zero_fraction,
+                v});
+          child_idx = parent_idx;
+          parent_idx = parent_id[child_idx];
+        }
+        // Root node has feature -1
+        {
+          int group_id = tree_idx % num_groups;
+          paths->paths.push_back(
+              gpu_treeshap::PathElement<MySplitCondition<ThresholdType>>{
+                path_idx, -1, group_id, {-inf, inf}, 1.0, v});
+          path_idx++;
+        }
+      }
+    }
+    tree_idx++;
+  }
+  paths->global_bias = model.param.global_bias;
+  paths->task_type = model.task_type;
+  paths->task_param = model.task_param;
+  paths->average_tree_output = model.average_tree_output;
+  paths->num_tree = static_cast<int>(model.trees.size());
+
+  return static_cast<ExtractedPathHandle>(path_container.release());
+}
+
+void extract_paths(ModelHandle model, ExtractedPathHandle* extracted_paths) {
+  const tl::Model& model_ref = *static_cast<tl::Model*>(model);
+
+  *extracted_paths = model_ref.Dispatch([&](const auto& model_inner) {
+    // model_inner is of the concrete type tl::ModelImpl<threshold_t, leaf_t>
+    return extract_paths_impl(model_inner);
+  });
+}
+
+void gpu_treeshap(ExtractedPathHandle extracted_paths, const float* data,
+                  std::size_t n_rows, std::size_t n_cols, float* out_preds) {
+  ExtractedPath& path_ref = *static_cast<ExtractedPath*>(extracted_paths);
+  DenseDatasetWrapper X(data, n_rows, n_cols);
+  std::size_t num_groups = 1;
+  if (path_ref.task_type == tl::TaskType::kMultiClfGrovePerClass
+      && path_ref.task_param.num_class > 1) {
+    num_groups = static_cast<std::size_t>(path_ref.task_param.num_class);
+  }
+  thrust::device_ptr<float> out_preds_ptr
+    = thrust::device_pointer_cast(out_preds);
+  path_ref.Dispatch([&](auto& paths) {
+    gpu_treeshap::GPUTreeShap(X, paths.paths.begin(), paths.paths.end(),
+                              num_groups, out_preds_ptr,
+                              out_preds_ptr
+                              + (n_rows * num_groups * (n_cols + 1)));
+  });
+  float global_bias = path_ref.global_bias;
+  int num_tree = path_ref.num_tree;
+  auto count_iter = thrust::make_counting_iterator(0);
+  if (path_ref.average_tree_output) {
+    thrust::for_each(thrust::device, count_iter,
+        count_iter + (n_rows * num_groups * (n_cols + 1)),
+      [=] __device__(std::size_t idx) {
+        out_preds[idx] /= num_tree;
+      });
+  }
+  thrust::for_each(thrust::device, count_iter,
+      count_iter + (n_rows * num_groups),
+    [=] __device__(std::size_t idx) {
+      out_preds[(idx + 1) * (n_cols + 1) - 1] += global_bias;
+    });
+}
+
+void free_extracted_paths(ExtractedPathHandle extracted_paths) {
+  delete static_cast<ExtractedPath*>(extracted_paths);
```

Comment:
It looks like this is a simple enough `free` function that an RAII implementation definitely makes sense.